### PR TITLE
Harmonize Cluster Warnings

### LINKED ISSUE #958: Harmonize Cluster Warnings
Currently we have several different components to show warnings on the cluster list and details page:

- SelfTerminationWarning
- VersionExpirationWarning
- ConstraintWarning (type maintenance / type hibernation)
- HibernationScheduleWarning

As it is implemented right now, it can happen that multiple (similar) icons appear for a cluster on the list. Moreover, some of the components show warnings for different types (as list in a popover), some show tooltips.

A solution could be to build one generic component for all warning types. We need to decide which icon(s) to show in this case.

### PR DESCRIPTION
**Pull request by grolu**
**What this PR does / why we need it**:
Introduce one generic component for all shoot warning types

**Which issue(s) this PR fixes**:
Fixes #958

**Special notes for your reviewer**:

**Release note**:

```bugfix user

```

### CODE REVIEW

File: frontend/src/components/ShootWarnings/ClusterExpirationMessage.vue [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+  components: {
+    TimeString
+  },
+  props: {
+    shootExpirationTimestamp: {
+      type: String,
+      required: true
+    }
+  },
+  methods: {
```
Comment (holgerkoser) [inline frontend/src/components/ShootWarnings/ClusterExpirationMessage.vue:30]:
```suggestion
  computed: {
```

File: frontend/src/components/ShootWarnings/ClusterExpirationMessage.vue [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+  </div>
+</template>
+
+<script>
+
+import { isValidTerminationDate } from '@/utils'
+import TimeString from '@/components/TimeString'
+
+export default {
+  name: 'ClusterExpirationMessage',
```
Comment (holgerkoser) [inline frontend/src/components/ShootWarnings/ClusterExpirationMessage.vue:20]:
```suggestion
```

File: frontend/src/components/ShootWarnings/ConstraintWarningMessage.vue [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+  <div>
+    <span class="font-weight-bold">{{constraintCaption}}: </span>
+    <span>{{constraintMessage}}</span>
+  </div>
+</template>
+
+<script>
+
+export default {
+  name: 'ConstraintWarningMessage',
```
Comment (holgerkoser) [inline frontend/src/components/ShootWarnings/ConstraintWarningMessage.vue:17]:
```suggestion
```

File: frontend/src/components/ShootWarnings/ConstraintWarningMessage.vue [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+  props: {
+    constraintCaption: {
+      type: String,
+      required: true
+    },
+    constraintMessage: {
+      type: String,
+      required: true
+    }
```
Comment (holgerkoser) [inline frontend/src/components/ShootWarnings/ConstraintWarningMessage.vue:26]:
```suggestion
    caption: {
      type: String,
      required: true
    },
    message: {
      type: String,
      required: true
    }
```

File: frontend/src/components/ShootWarnings/ConstraintWarningMessage.vue [hunk 5]
```diff
@@ -0,0 +1,29 @@
+<!--
+SPDX-FileCopyrightText: 2021 SAP SE or an SAP affiliate company and Gardener contributors
+
+SPDX-License-Identifier: Apache-2.0
+-->
+
+<template>
+  <div>
+    <span class="font-weight-bold">{{constraintCaption}}: </span>
+    <span>{{constraintMessage}}</span>
```
Comment (holgerkoser) [inline frontend/src/components/ShootWarnings/ConstraintWarningMessage.vue:10]:
```suggestion
    <span class="font-weight-bold">{{caption}}: </span>
    <span>{{message}}</span>
```

File: frontend/src/components/ShootWarnings/ConstraintWarningMessage.vue [hunk 6]
```diff
@@ -0,0 +1,29 @@
+<!--
```
Comment (holgerkoser) [inline frontend/src/components/ShootWarnings/ConstraintWarningMessage.vue:1]: Rename the the file to `ConstraintMessage.vue` or even more generic

File: frontend/src/components/ShootWarnings/K8sExpirationMessage.vue [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+        <template v-slot:activator="{ on }">
+          <span class="font-weight-bold" v-on="on"><time-string :date-time="expirationDate" mode="future"></time-string></span>
+        </template>
+        <span>{{getDateFormatted(expirationDate)}}</span>
+      </v-tooltip>
+      <span>. </span>
+    </span>
+    <span v-else>Kubernetes version of this cluster is expired. </span>
+    <span v-if="severity==='info'">Version will be updated in the next maintenance window</span>
+    <template v-if="severity==='warning'">
```
Comment (holgerkoser) [inline frontend/src/components/ShootWarnings/K8sExpirationMessage.vue:20]:
```suggestion
    <template v-else-if="severity === 'warning'">
```

File: frontend/src/components/ShootWarnings/K8sExpirationMessage.vue [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+      <v-tooltip right>
+        <template v-slot:activator="{ on }">
+          <span class="font-weight-bold" v-on="on"><time-string :date-time="expirationDate" mode="future"></time-string></span>
+        </template>
+        <span>{{getDateFormatted(expirationDate)}}</span>
+      </v-tooltip>
+      <span>. </span>
+    </span>
+    <span v-else>Kubernetes version of this cluster is expired. </span>
+    <span v-if="severity==='info'">Version will be updated in the next maintenance window</span>
```
Comment (holgerkoser) [inline frontend/src/components/ShootWarnings/K8sExpirationMessage.vue:19]:
```suggestion
    <span v-if="severity === 'info'">Version will be updated in the next maintenance window</span>
```

File: frontend/src/components/ShootWarnings/K8sExpirationMessage.vue [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+    <template v-if="severity==='warning'">
+      <span v-if="isValidTerminationDate">Version update will be enforced after that date</span>
+      <span v-else>Version update will be enforced soon</span>
+    </template>
+    <span v-if="severity==='error'">Version update not possible due to missing update path. Please contact your landscape administrator</span>
```
Comment (holgerkoser) [inline frontend/src/components/ShootWarnings/K8sExpirationMessage.vue:24]:
```suggestion
    <span v-else-if="severity === 'error'">Version update not possible due to missing update path. Please contact your landscape administrator</span>
```

File: frontend/src/components/ShootWarnings/K8sExpirationMessage.vue [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+  </div>
+</template>
+
+<script>
+
+import { getDateFormatted } from '@/utils'
+import TimeString from '@/components/TimeString'
+
+export default {
+  name: 'K8sExpirationMessage',
```
Comment (holgerkoser) [inline frontend/src/components/ShootWarnings/K8sExpirationMessage.vue:34]:
```suggestion
```

File: frontend/src/components/ShootWarnings/K8sExpirationMessage.vue [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+  <div>
+    <span v-if="isValidTerminationDate">Kubernetes version of this cluster expires
+      <v-tooltip right>
+        <template v-slot:activator="{ on }">
+          <span class="font-weight-bold" v-on="on"><time-string :date-time="expirationDate" mode="future"></time-string></span>
+        </template>
+        <span>{{getDateFormatted(expirationDate)}}</span>
+      </v-tooltip>
+      <span>. </span>
+    </span>
```
Comment (holgerkoser) [inline frontend/src/components/ShootWarnings/K8sExpirationMessage.vue:17]:
```suggestion
    <span v-if="isValidTerminationDate">Kubernetes version of this cluster expires
      <v-tooltip right>
        <template v-slot:activator="{ on }">
          <time-string v-on="on" :date-time="expirationDate" mode="future" class="font-weight-bold"></time-string>
        </template>
        {{getDateFormatted(expirationDate)}}
      </v-tooltip>. </span>
```

File: frontend/src/components/ShootWarnings/NoHibernationScheduleMessage.vue [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+      <router-link :to="{ name: 'ShootItemHibernationSettings', params: { name: shootName, namespace: shootNamespace } }">configure</router-link>
+      a hibernation schedule or explicitly deactivate scheduled hibernation for this cluster.
+    </template>
+  </div>
+</template>
+
+<script>
+
+export default {
+  name: 'NoHibernationScheduleMessage',
```
Comment (holgerkoser) [inline frontend/src/components/ShootWarnings/NoHibernationScheduleMessage.vue:21]:
```suggestion
```

File: frontend/src/components/ShootWarnings/ShootWarnings.vue [hunk 13]
```diff
@@ -0,0 +1,327 @@
...
+    visible () {
+      return !this.isShootMarkedForDeletion && (
+        this.isK8sWarning ||
+        this.isMachineImageWarning ||
+        this.isNoHibernationScheduleWarning ||
+        this.isClusterExpirationWarning ||
+        this.isConstraintWarning
+      )
+    },
+    warningItems () {
```
Comment (holgerkoser) [inline frontend/src/components/ShootWarnings/ShootWarnings.vue:121]: What is the difference between a warning which is a warning and a warning which is an error or and just an info. That's why I named it messages. It is unclear to me what we call message and what we call warning and how an error could be a warning.

File: frontend/src/components/ShootWarnings/ShootWarnings.vue [hunk 14]
```diff
@@ -0,0 +1,327 @@
...
+          })
+        })
+      }
+      if (this.isNoHibernationScheduleWarning) {
+        items.push({
+          key: 'noHibernationSchedule',
+          icon: 'mdi-calendar-alert',
+          color: 'info',
+          messageComponent: 'NoHibernationScheduleMessage',
+          messageComponentProperties: {
```
Comment (holgerkoser) [inline frontend/src/components/ShootWarnings/ShootWarnings.vue:161]: Why not
```js
component: {
  name: 'kebap-case-name',
  props: {
    ...
  }
}
```

File: frontend/src/components/ShootWarnings/ShootWarnings.vue [hunk 15]
```diff
@@ -0,0 +1,327 @@
...
+            }
+          })
+        })
+      }
+      if (this.isNoHibernationScheduleWarning) {
+        items.push({
+          key: 'noHibernationSchedule',
+          icon: 'mdi-calendar-alert',
+          color: 'info',
+          messageComponent: 'NoHibernationScheduleMessage',
```
Comment (holgerkoser) [inline frontend/src/components/ShootWarnings/ShootWarnings.vue:160]: component names should be spcified in the form case as in the html

File: frontend/src/components/ShootWarnings/ShootWarnings.vue [hunk 16]
```diff
@@ -0,0 +1,327 @@
...
+            messageComponent: 'WorkerGroupExpirationMessage',
+            messageComponentProperties: {
+              expirationDate,
+              isValidTerminationDate,
+              severity,
+              name,
+              workerName,
+              version
+            }
+          })
```
Comment (holgerkoser) [inline frontend/src/components/ShootWarnings/ShootWarnings.vue:152]:
```suggestion
        forEach(this.expiredWorkerGroups, workerGroup => {
          items.push({
            key: `image_${workerName}_${name}`,
            icon: 'mdi-update',
            color: this.colorForSeverity(severity),
            messageComponent: 'WorkerGroupExpirationMessage',
            messageComponentProperties: { ...workerGroup }
          })
```

File: frontend/src/components/ShootWarnings/WorkerGroupExpirationMessage.vue [hunk 17]
```diff
@@ -0,0 +1,71 @@
...
+  </div>
+</template>
+
+<script>
+
+import { getDateFormatted } from '@/utils'
+import TimeString from '@/components/TimeString'
+
+export default {
+  name: 'WorkerGroupExpirationMessage.',
```
Comment (holgerkoser) [inline frontend/src/components/ShootWarnings/WorkerGroupExpirationMessage.vue:35]:
```suggestion
```

File: frontend/src/components/ShootWarnings/WorkerGroupExpirationMessage.vue [hunk 18]
```diff
@@ -0,0 +1,71 @@
...
+SPDX-License-Identifier: Apache-2.0
+-->
+
+<template>
+  <div>
+    <span>Machine image <span class="font-weight-bold">{{name}} | Version: {{version}}</span> of worker group <span class="font-weight-bold">{{workerName}} </span></span>
+    <span v-if="isValidTerminationDate">expires
+      <v-tooltip right>
+        <template v-slot:activator="{ on }">
+          <span class="font-weight-bold" v-on="on"><time-string :date-time="expirationDate" mode="future"></time-string></span>
```
Comment (holgerkoser) [inline frontend/src/components/ShootWarnings/WorkerGroupExpirationMessage.vue:13]:
```suggestion
          <time-string v-on="on" :date-time="expirationDate" mode="future" class="font-weight-bold"></time-string>
```

File: frontend/src/components/ShootWarnings/WorkerGroupExpirationMessage.vue [hunk 19]
```diff
@@ -0,0 +1,71 @@
...
+        </template>
+        <span>{{getDateFormatted(expirationDate)}}</span>
```
Comment (holgerkoser) [inline frontend/src/components/ShootWarnings/WorkerGroupExpirationMessage.vue:15]:
```suggestion
       {{getDateFormatted(expirationDate)}}
```

File: frontend/src/components/ShootWarnings/ShootWarnings.vue (hunk 14 above)
Comment (grolu) [reply frontend/src/components/ShootWarnings/ShootWarnings.vue:161]: use kebap case and hierarchy

File: frontend/src/components/ShootWarnings/ConstraintWarningMessage.vue (hunk 6 above)
Comment (grolu) [reply frontend/src/components/ShootWarnings/ConstraintWarningMessage.vue:1]: done

File: frontend/src/components/ShootWarnings/ShootWarnings.vue (hunk 13 above)
Comment (grolu) [reply frontend/src/components/ShootWarnings/ShootWarnings.vue:121]: done, do not use term _warning_ for component and generic methods as this is misleading. Use term _message_

File: frontend/src/components/ShootWarnings/ShootWarnings.vue (hunk 14 above)
Comment (grolu) [reply frontend/src/components/ShootWarnings/ShootWarnings.vue:161]: done

File: frontend/src/components/ShootWarnings/ShootWarnings.vue (hunk 15 above)
Comment (grolu) [reply frontend/src/components/ShootWarnings/ShootWarnings.vue:160]: done

File: frontend/src/components/ShootMessages/ShootMessages.vue [hunk 20]
```diff
@@ -0,0 +1,301 @@
...
+    margin: 0px !important;
+  }
+
+  ::v-deep .v-list-item__content {
+    padding: 0px;
+  }
+
+  ::v-deep .v-list-item {
+    min-height: 0px;
+  }
```
[the remark this answers is not in the document]
Comment (grolu) [reply frontend/src/components/ShootMessages/ShootMessages.vue:311]: wrapped with .v-card to avoid multiple usage of ::v-deep

File: frontend/src/components/ShootMessages/WorkerGroupExpirationMessage.vue [hunk 21]
```diff
@@ -0,0 +1,70 @@
...
+      </v-tooltip>
+      <span>. </span>
+    </span>
+    <span v-else>is expired. </span>
+    <span v-if="severity==='info'">Version will be updated in the next maintenance window</span>
+    <template v-if="severity==='warning'">
+      <span v-if="isValidTerminationDate">Machine Image update will be enforced after that date</span>
+        <span v-else>Machine Image update will be enforced soon</span>
+    </template>
+    <span v-if="severity==='error'">Machine Image update not possible as no newer version is available. Please choose another operating system</span>
```
[the remark this answers is not in the document]
Comment (grolu) [reply frontend/src/components/ShootMessages/WorkerGroupExpirationMessage.vue:25]: done

File: frontend/src/mixins/shootItem.js [hunk 22]
```diff
@@ -132,6 +132,9 @@ export const shootItem = {
     shootDomain () {
       return get(this.shootSpec, 'dns.domain')
     },
+    shootDnsProviders () {
```
Comment (holgerkoser) [inline frontend/src/mixins/shootItem.js:135]: I think this should not be part of this PR. Correct me if I am wrong.

Comment (grolu) [reply frontend/src/mixins/shootItem.js]: correct, I removed it

**Review by holgerkoser — APPROVED** [review]
/lgtm
